core/src/main/scala/org/apache/spark/util/AccumulatorV2.scala

```diff
@@ -444,7 +444,7 @@ class CollectionAccumulator[T] extends AccumulatorV2[T, java.util.List[T]] {
 
   override def copy(): CollectionAccumulator[T] = {
     val newAcc = new CollectionAccumulator[T]
-    newAcc._list.addAll(_list)
+    newAcc.merge(this)
```

Comment:
I don't have a strong opinion so let me change `merge()` with `synchronized` block as you suggested. 

I also propose changing `BlockStatusesAccumulator` to use `setValue()` rather than `merge()` to avoid extra type matching.